### ISSUE TITLE
Recursive `RlpStream::append` in `Encodable`/`RlpEncodable` leads to invalid list counts

### ISSUE THREAD

**Issue by maciejhirsz** [ethcore/parity#4356]
[Some types](https://github.com/ethcore/parity/blob/8404edb656693270ec1cd3956d204f625d28ec7c/ethcore/src/engines/tendermint/message.rs#L163-L167) implement the `Encodable` trait as follows:

```rust
impl Encodable for Step {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.append(&self.number());
	}
}
```

This seems to generally work, but incorrectly assumes two items got appended when, in fact, only one has been. Isolated way to reproduce the problem:

```rust
    #[test]
    fn message_info_rlp_step() {
        let mut s = RlpStream::new_list(2);
        s.append(&Step::Precommit);
        assert!(!s.is_finished());
    }
```

The stream becomes marked as finished even though only one item has been appended. This is due to the double (recursive) call of `RlpStream::append` - once when appending the item, and another one in the implementation of the trait.

Following through in `RlpStream`:
```rust
	fn append<'a, E>(&'a mut self, value: &E) -> &'a mut Self where E: RlpEncodable {
		self.finished_list = false;
		value.rlp_append(self); // `E` calls append again
		if !self.finished_list {
			self.note_appended(1); // if the list isn't finished, `note_appended` will be called again
		}
		self
	}
```

## Solution:
+ ~Make `RlpStream::append_value` public.~
+ ~The `Encodable` implementations that aren't creating sub-lists should use `append_value` instead of `append`.~
+ Instead of `s.append(&self.to_another_type())` call `self.to_another_type().rlp_append()` to avoid double `RlpStream::append` calls.

## Food for thought:
+ This can't be tested with happy path testing since, should the list be finished after the first `note_appended(1)`, all tests will be green even though two attempts to increase the count has been made. Adding negative tests for `RlpStream::is_finished` would solve that.
* Should `RlpStream` panic :frowning: on overflows? There is no error handling except in `UntrustedRlp`.

**Comment by arkpar** [comment]
Maybe just remove `note_appended` from `append` and call it from `append_value` instead? @debris what do you think?

**Comment by arkpar** [comment]
Overflowing `RlpStream` is a programming error. You'd likely to run OOM first. So I guess it is ok to panic here.

**Comment by maciejhirsz** [comment]
`note_appended` already has a built in panic for overflows, but it's never triggered under normal conditions since `append` does not call `note_appended` when overflow happens. It seems to me that removing the check and forcing `note_appended` to run it's course and panic is the simplest solution.

That said, I tried that already, but a bunch of tests turned red. I didn't feel too confident about my understanding of why the boolean flag was there to begin with, so I left it in place for now.

**Comment by debris** [comment]
> Should RlpStream panic 😦 on overflows? There is no error handling except in UntrustedRlp.

Yes, I agree with @arkpar here

> That said, I tried that already, and a bunch of tests turned red. I didn't feel too confident about my understanding of why the boolean flag was there to begin with, so I left it in place for now.

It's because, it's allowed to append raw items to rlp stream, without declaring a list in a first place

-------------------------------

Now a little about a problem.

```rust
impl Encodable for Step {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.append(&self.number());
	}
}
```

If we compare above implementation with other implementations in parity, we can notice that:

1. `s.append(..)` never exists on it's own inside `rlp_append`
2. if it exist inside `rlp_append` function, it's usually prepended with `s.begin_list(..)` call. This guarantees that new items are appended as one element - list
3. there is only one way of appending publicly visible items `s.append(&wrapper_structure.as_raw_struct())`

In your case, I would use `s.append(&wrapper_structure.as_raw_struct())` for now. Generally, I opt for refactoring rlp module in nearest future (after 1.6?) and I think it would be best to come back to this discussion then

**Comment by maciejhirsz** [comment]
> there is only one way of appending publicly visible items `s.append(&wrapper_structure.as_raw_struct())`

That sounds volatile. Delegating wrapper type `Encodable::rlp_append` to the inner type `Encodable::rlp_append` works well enough with clearer intent:

```rust
impl Encodable for Step {
    fn rlp_append(&self, s: &mut RlpStream) {
        RlpEncodable::rlp_append(&self.number(), s); // number() returns u8
    }
}
```

Considering the bug seems to be fixed for now (as soon as #4362 is merged), I assume the issue can be closed till we revisit RLP?
